packages/plugin-ext/src/main/browser/webview/webview.ts

```diff
@@ -0,0 +1,212 @@
+/********************************************************************************
+ * Copyright (C) 2018 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { BaseWidget } from '@theia/core/lib/browser/widgets/widget';
+import { IdGenerator } from '../../../common/id-generator';
+import { MiniBrowser } from '@theia/mini-browser/lib/browser/mini-browser';
+import { Disposable, DisposableCollection } from '@theia/core';
+
+export interface WebviewWidgetOptions {
+    readonly allowScripts?: boolean;
+}
+
+export interface WebviewEvents {
+    onMessage?(message: any): void;
+    onKeyboardEvent?(e: KeyboardEvent): void;
+    onLoad?(contentDocument: Document): void;
+}
+
+export class WebviewWidget extends BaseWidget {
+    private static readonly ID = new IdGenerator('webview-widget-');
+
+    protected readonly toDispose = new DisposableCollection();
+    private iframe: HTMLIFrameElement;
+    private state: string | undefined = undefined;
+    private loadTimeout: number | undefined;
+    //    private pendingMessages
+    constructor(title: string, private options: WebviewWidgetOptions, private eventDelegate: WebviewEvents) {
+        super();
+        this.id = WebviewWidget.ID.nextId();
+        this.title.closable = true;
+        // this.title.caption = this.title.label = this.props.name || 'Browser';
+        this.title.label = title;
+        // this.title.iconClass = this.props.iconClass || MiniBrowser.ICON;
+        this.addClass(MiniBrowser.Styles.MINI_BROWSER);
+
+    }
+
+    protected onFrameLoad(): void {
+        // this.hideLoadIndicator();
+        // this.focus();
+    }
+
+    protected handleMessage(message: any) {
+        switch (message.command) {
+            case 'onmessage':
+                this.eventDelegate.onMessage!(message.data);
+                break;
+            case 'do-update-state':
+                this.state = message.data;
+        }
+    }
+
+    postMessage(message: any) {
+        this.iframe.contentWindow!.postMessage(message, '*');
+    }
+
+    setHTML(html: string) {
+        html = html.replace('theia-resource:/', '/webview/');
+        html = html.replace('vscode-resource:/', '/webview/');
```

Comment:
`vscode-resource:/` handling should be provided by `plugin-ext-vscode` extension

Reply:
Ok

Reply:
https://github.com/theia-ide/theia/pull/3484/commits/a563e22884567ddc3ebe72233a446ea6c3290c8b#diff-66cd2825aa616ad9a26b1f62e1b21e5bR60